### ISSUE TITLE
401 status from server not being propagated to CBLReplication.lastError

### ISSUE THREAD

**Issue by snej** [couchbase/couchbase-lite-ios#900]
A [report from the forum](https://forums.couchbase.com/t/sync-gateway-returning-401-unauthorised-does-not-set-lasterror/5027):

>I've noticed that if the user password changes in CB, a device that is syncing logs in the console (repeatedly):
```
2015-09-01 17:26:05.597 KetoDiet[1767:480216] CBLMultipartDownloader[/Cloud/Sync.ashx/ketodiet/CMI_BEAB9D51F53945D89904A69A101D7AC3]: Got error Error Domain=CBLHTTP Code=401 "401 unauthorized" UserInfo=0x174e74080 {NSURL=http://192.168.0.95/Cloud/Sync.ashx/ketodiet/CMI_BEAB9D51F53945D89904A69A101D7AC3?rev=2-9cf168d5d3a4ddaf316073d05065d3d3&revs=true&attachments=true, NSLocalizedFailureReason=unauthorized, NSLocalizedDescription=401 unauthorized}
```
>but, it never sets the pull or push lastError field...? I'm checking for these fields in my kCBLReplicationChangeNotification handler. They are both (push.lastError / pull.lastError) set to nil despite the 401 error returned from the server.

**Comment by snej** [comment]
The problem is that -[CBLRestPuller pullRemoteRevision:] handles a response error by calling `[strongSelf revision: rev failedWithError: error]`. This is appropriate for a document-level error like a 403 or 404, but not for a connection-level error like 401 — all it does is mark the revision as failed and keep going.

This is actually a regression introduced a year ago in 87f3261 as part of the fix for #377. That fix made document-level errors from this HTTP call work correctly, but it broke connection-level errors like this one.

The right fix seems to be to look at the error and treat the two types differently. I think the only document-level errors possible here are 403 and 404. Maybe 410, which is pretty much synonymous with 404 although I don't think either CouchDB or Sync Gateway ever returns it.

**Comment by snej** [comment]
@hideki: You should check whether Android has the same problem, because #377 was a copy of https://github.com/couchbase/couchbase-lite-android/issues/376 and I think we fixed both of them the same way.
